### PR TITLE
[New Feature] integrate Liger Kernel with fused-CE for NPU + fix conv1d/seed bugs

### PR DESCRIPTION
**Pull request by ys2025-AI**
# PR type
- [x] Bug Fix
- [x] New Feature
- [x] Document Updates
- [ ] More Models or Datasets Support

# PR information

Liger Kernel integration:
- Add liger_builtin() bundle with _prefer_cann_on_npu() that swaps Liger
  Triton-on-Ascend kernels for CANN equivalents (faster on NPU)
- Add LigerFusedLinearCrossEntropyLoss + TransformersFusedCEPatch that
  skips lm_head GEMM and [B,T,V] logits materialisation
- Wire fused-CE in fsdp2.py/sp_fsdp_dense.py/ep_fsdp2_lora_qwen3_5_moe.py
  (task='fused_lm_ce' + set_loss + pre-loop barrier)
- Add --enable-liger / --no-enable-fused-ce CLI flags

Bug fixes:
- causal_conv1d.py: fix npu_causal_conv1d_fn axis confusion when
  conv_dim == seq_len (e.g. 35B-A3B with max_length=8192); add
  x.shape[1] != D check to disambiguate [B,D,T] vs [B,T,D] layouts
- framework.py: fix torch.manual_seed gated behind is_gpu_available()
  on NPU-only hosts; move to unconditional + independent NPU branch

Tests:
- test_qwen35_fla_bwd_precision.py: FLA on/off backward precision
- test_liger_builtin.py, test_liger_fused_linear_ce.py, test_fused_ce_patch.py

## Experiment results

Benchmark (Qwen3.5-35B-A3B, 4x910B3, max_len=8192):
- fused-CE saves 12-20 GB/NPU (20-30%) across all fsdp2，sp+fsdp2, ep+fsdp2
- Enables EP batch=8 (was OOM without fused-CE)
- SP path 1.6x speedup from skipping logits SP all-gather

### CODE REVIEW

File: cookbook/transformers/ep_fsdp2_lora_qwen3_5_moe.py [hunk 1]
```diff
@@ -116,10 +136,20 @@ def train():
...
+    if _use_fused_ce:
+        import torch.distributed as _dist
+        if _dist.is_available() and _dist.is_initialized():
+            _dist.barrier()
+        try:
+            import torch_npu
+            if torch_npu.npu.is_available():
+                torch_npu.npu.synchronize()
+        except ImportError:
+            pass
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply cookbook/transformers/ep_fsdp2_lora_qwen3_5_moe.py]: accepted

File: cookbook/transformers/fsdp2.py [hunk 2]
```diff
@@ -97,8 +129,37 @@ def train():
...
+            if torch_npu.npu.is_available():
+                torch_npu.npu.synchronize()
+        except ImportError:
+            pass
+        try:
+            import torch
+            if torch.cuda.is_available():
+                torch.cuda.synchronize()
+        except Exception:
+            pass
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply cookbook/transformers/fsdp2.py]: accepted

File: src/twinkle/loss/liger_fused_linear_cross_entropy.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        self._warned = False
+
+    def _materialise_ce(self, inputs, outputs, lm_head, hidden, **kwargs):
+        """Materialise logits from the stashed lm_head weight (NOT its forward,
+        which is identity under the patch) and run standard cross-entropy."""
+        from .cross_entropy import CrossEntropyLoss
+        weight = lm_head.weight
+        if hasattr(weight, 'full_tensor'):
+            weight = weight.full_tensor()
+        bias = getattr(lm_head, 'bias', None)
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/loss/liger_fused_linear_cross_entropy.py:152]: accepted

File: cookbook/transformers/fsdp2.py [hunk 4]
```diff
@@ -56,10 +59,36 @@ def train():
...
-    # npu patch
-    if Torch.is_npu_available():
-        model = kernelize(model, npu_builtin(model))
+    # Discover the actual decoder-layer class name(s) from the live model — the
+    # ``model_type.title() + 'DecoderLayer'`` heuristic misnames MoE models
+    # (e.g. ``qwen3_5_moe`` -> ``Qwen3_5_MoeDecoderLayer`` vs the real
+    # ``Qwen3_5MoeDecoderLayer``).
+    discovered = {type(m).__name__ for m in model.model.modules()
+                  if type(m).__name__.endswith('DecoderLayer')}
+    model.model._no_split_modules = discovered or {model.model.config.model_type.title() + 'DecoderLayer'}
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply cookbook/transformers/fsdp2.py]: accepted

File: cookbook/transformers/sp_fsdp_dense.py [hunk 5]
```diff
@@ -81,11 +94,26 @@ def train():
...
+    if _use_fused_ce:
+        import torch.distributed as _dist
+        if _dist.is_available() and _dist.is_initialized():
+            _dist.barrier()
+        try:
+            import torch_npu
+            if torch_npu.npu.is_available():
+                torch_npu.npu.synchronize()
+        except ImportError:
+            pass
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply cookbook/transformers/sp_fsdp_dense.py]: accepted

File: src/twinkle/kernel/causal_conv1d.py [hunk 6]
```diff
@@ -1093,10 +1093,21 @@ def npu_causal_conv1d_fn(
...
+    D, W = weight.shape[0], weight.shape[1]
+    is_bt_d_layout = (
+        x.dim() == 3 and weight.dim() == 2 and x.shape[-1] == D  # last dim is the channel dim
+        and x.shape[1] != D  # second dim is NOT D -> genuinely [B, T, D]
+        and D != W  # sanity: D != kernel_size
+    )
+    if is_bt_d_layout:
         weight_t = weight.transpose(0, 1).contiguous()
         y_t, _ = causal_conv1d(x=x, weight=weight_t, bias=bias, activation=activation, cu_seqlens=cu_seqlens)
         return y_t
```
[the remark this answers is not in the document]
Comment (ys2025-AI) [reply src/twinkle/kernel/causal_conv1d.py:1113]: False positive — the `else` branch exists at line 1114 of `causal_conv1d.py`:

```python
    if is_bt_d_layout:
        weight_t = weight.transpose(0, 1).contiguous()
        y_t, _ = causal_conv1d(x=x, weight=weight_t, ...)
        return y_t
    else:
        x_t = x.transpose(1, 2).contiguous()
        weight_t = weight.transpose(0, 1).contiguous()
        y_t, _ = causal_conv1d(x=x_t, weight=weight_t, ...)
        return y_t.transpose(1, 2).contiguous()
```

The diff snippet shown in the review only displayed lines 1110-1113 (the `if` block), so the `else` block at 1114-1118 was not visible. The standard `[B, D, T]` path is fully handled — no implicit `None` return.

**Review by tpx818 — APPROVED** [review]
